### PR TITLE
Closes #9391 Add support for geckoView site permissions storage.

### LINKED ISSUE #9391: Migrate site permission db to new gv api
The gv team is working on a new api for site permissions right now we are storing all the data related in our own db but the new api will require us to migrate the data over as the new api offers storage.

https://bugzilla.mozilla.org/show_bug.cgi?id=1654832

### PR DESCRIPTION
**Pull request by Amejia481**
[Abstract document](https://docs.google.com/document/d/1KUq0gejnFm5erkHNkswm8JsT7nLOmWvs1KEGFz9FWxk/edit#heading=h.ls1dr18v7zrx)

Fenix related PR https://github.com/mozilla-mobile/android-components/pull/10176

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionsMigration.kt
Comment (Amejia481) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionsMigration.kt:24]: The main idea is to call `performMigration` on Fenix startup, this will take some time depending on the amount of user data but it will be done once, also we are going to block when a permission request arrive until the migration is finished.

File: components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
Comment (Amejia481) [inline components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt:367]: We will need to coordinate to not duplicate migration data.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionsMigration.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionsMigration.kt:24]: I think it will be much easier if we only have one blocking migration, instead doing `async` and have to synchronized multiple migrations at the time.

**Comment by Amejia481** [comment]
@pocmo I chatted with @csadilek, we were thinking about changing the initial approach of of doing one shot migration. As there are some new information that we were not aware before.  The new GV API will partially store some types of permissions like notification, location, and autoplay but for other types like camera and microphone, ac will still need to store them, as those are not going to change.

Due to we still have to keep the previous permission storage, we are thinking about doing an incremental migration where every time an user queries a site permission, we check in the old storage, we delete it from the old storage and add it to the new one. That will be a bit safer and won't required user to do a full migration, for sites that they don't use. What do you think?

**Comment by pocmo** [comment]
> Due to we still have to keep the previous permission storage, we are thinking about doing an incremental migration where every time an user queries a site permission, we check in the old storage, we delete it from the old storage and add it to the new one. That will be a bit safer and won't required user to do a full migration, for sites that they don't use. What do you think?

Sounds good to me. This also doesn't block us from doing a full migration at a later time if we decide to do that. 👍

**Comment by pocmo** [comment]
In addition to that I wonder if we could report the number of permissions in the legacy storage via telemetry? We should see this number go down over time and maybe at some point we wouldn't need to do any migration at all? :)

**Comment by Amejia481** [comment]
> In addition to that I wonder if we could report the number of permissions in the legacy storage via telemetry? We should see this number go down over time and maybe at some point we wouldn't need to do any migration at all? :)

That sounds like a good idea 👍🏽

**Comment by Amejia481** [comment]
Apart from the code review it will be ideal if we could do a round of manual tests to discovered possible issues early.

Some tools for manual testing: 
* Website to simulate site permissions request https://permission.site/ 
*  Related Fenix PR https://github.com/mozilla-mobile/fenix/pull/19553/ 
* [The Database Inspector In Android studio](https://developer.android.com/studio/inspect/database)
* Site permission status DB values `BLOCKED(-1), NO_DECISION(0), ALLOWED(1)`

I think we could cover the follow test cases:

1) With a clean install check you are able to allow/deny permissions.
When you check the checkbox allowing/denying permission must be persistented,  content permissions **(Notification,Location, Media key access, local storage, auto play Audible, and auto play in Audible)** should not be store in the db while media  permissions (**microphone and camera**) should be store in the DB.

2) After updating permissions (In the doorhanger) content permission should now be in the DB while media permission should be updated in the DB.

3) Updating permissions from the Settings -> Site permissions -> Exceptions page should update the permission on the next request, content permissions should not be updated in the DB while media permissions should be.

4) Removing individual permissions from the Settings -> Site permissions -> Exceptions page, clear any permissions.

5) Removing all permissions from the Settings -> Site permissions -> Exceptions page, clear all permissions.

6) Testing temporary site permissions that happen when permissions are allowed/granted without checking the checkbox, 
only location, camera and microphone have the checkbox.  temporary permissions are not persistent and get cleared out when reloading the page or navigating away.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (Amejia481) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:165]: @pocmo I remembered you mentioned the possibility of add some telemetry for when permissions get migrated? As permissions get migrated little by little, do you think is there any additional that we should capture?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:454]: What's the plan for this going forward? I assume we will need to pass those permissions along to be able to display them in the UI [but we currently aren't?]?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt:43]: There may be a lot of individual permission migrations happening. I wonder if it needs to be this fine-grained? Maybe just having a "not migrated permissions" count per user is good enough and shows us how far along we are in the migration?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt:78]: What's the reason for this `synchronized` block here? 🤔

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:60]: Is `onDiskStorage` our own storage implementation? Do we still need to save changes here? I thought from now on we basically only need to query our own storage (and migrate) and delete?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:78]: Same question as for `save`

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:84]: Ah, this may be the answer to my other question. There are still some left we need to save ourselves..?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:94]: Do we need this paging API in Fenix?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (csadilek) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:165]: Let's make sure to keep actually sending the telemetry in Fenix separate though so we don't have to wait for data review.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:133]: nit: "~~is~~ exists"

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt
Comment (csadilek) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:454]: I remember we talked about this before. Shouldn't we do something with `geckoPermissions` here. It seems like a bug to ignore that parameter. At a minimum, let's add a comment, please :).

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:56]: Temporary permissions are permissions where the user didn't check "remember my decision"?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:84]: Is there anything planned for the future?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt
Comment (csadilek) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt:76]: Hm don't you want the read and writes synchronized here? I am not sure why though? `grant` and `reject` should be called on the main thread only, right? What else is calling `isComplete`?

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/permission/SitePermissionsStorage.kt
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/permission/SitePermissionsStorage.kt:21]: nit: Saves

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt:43]: For doing that we will need to do a spesific query to extract that information, that is not part of the normal user flow,
 maybe we could take the telemetry part as follow up? An created a spesific that provide us that status. We could query that status and report it on startup.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/permission/SitePermissionsStorage.kt
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/permission/SitePermissionsStorage.kt:48]: nit: newline.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/permission/SitePermissionsStorage.kt
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/permission/SitePermissionsStorage.kt:54]: nit: Provides or better Returns?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (csadilek) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:60]: Shouldn't this be conditional somehow? Only save ourselves what we need to save i.e. what wasn't migrated or gecko doesn't handle or am I missing something?

**Review by csadilek — COMMENTED** [review]
Looks good, I had a few questions inline :)

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:78]: See https://github.com/mozilla-mobile/android-components/pull/10176#discussion_r635313893

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:84]: As far as I know far GV doesn't plan store media permissions in the near feature :(

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:94]: Yes, we need it for showing the list of permissions exceptions see https://github.com/mozilla-mobile/fenix/pull/19553/files#diff-d081cee3926a40933f15182c8a61ecf824345e867e4e0499624efd3d131ebf9fR62

 **Settings -> Site permissions -> Exceptions page**
image

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:56]: Yes, maybe I could add it on the comment above.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:454]: Every time that we navigate to site they are passing us the gv permissions for the site, I'm not sure how we could integrate it. As we are querying the store when it's needed and I don't know how useful it's this info at that spesific moment. I'll add a comment for it.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:60]: For saving new permissions we always favor GV, the conditional part it's being done in [updateGeckoPermissionIfNeeded](https://github.com/mozilla-mobile/android-components/pull/10176/files#diff-180aaba90cde778fa9e9cd1ff9e79849e25c8a3e3989947ed0e771e32560c805R150). 

We are only able to store on GV permissions that were previously requested, in this case we can only store the requested permission (`request`) in GV and the other values are store in our store until the user requested other values.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionRequest.kt:76]: You are right, my bad. I will remove it. We are only `granting` or `rejecting` on the main thread, and either way if we complete the gecko result in another thread it will be likely to throw an exception because of the looper issue.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt
Comment (pocmo) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt:43]: Yeah, +1: Let's do telemetry as a follow-up. We can add this "count" at any time and it wouldn't need to ship in the first version.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (pocmo) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:84]: Bummer! Maybe we can file it and see if they would be open to it? :)

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt
Comment (Amejia481) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoPermissionMigrationFacts.kt:43]: I filed https://github.com/mozilla-mobile/android-components/issues/10296 for it

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt
Comment (csadilek) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt:60]: Ok, got it!

**Comment by Amejia481** [comment]
@jonalmeida  discovered a bug while doing manual testing:


```
- Go to https://gauntface.github.io/simple-push-demo/
- Allow notification settings for the site.
- Send a push message and confirm you receive one.
- Click on security icon and disable notification permission.
- Send another push message.
The last push message should not go through.

```


So far, I found the issue is similar to what we saw on https://github.com/mozilla-mobile/android-components/issues/6322, permissions getting cached by a  service worker, in that case the workaround recommended for the GV team was to [clear the permissions cache, before updating](https://bugzilla.mozilla.org/show_bug.cgi?id=1618632), I did that but I'm still seeing the value to be cached by the GV storage, I will chat with the GV team to see if they can help or if we will need to file a new ticket for this.

**Comment by Amejia481** [comment]
> @jonalmeida discovered a bug while doing manual testing:
> 
> ```
> - Go to https://gauntface.github.io/simple-push-demo/
> - Allow notification settings for the site.
> - Send a push message and confirm you receive one.
> - Click on security icon and disable notification permission.
> - Send another push message.
> The last push message should not go through.
> ```
> 
> So far, I found the issue is similar to what we saw on #6322, permissions getting cached by a service worker, in that case the workaround recommended for the GV team was to [clear the permissions cache, before updating](https://bugzilla.mozilla.org/show_bug.cgi?id=1618632), I did that but I'm still seeing the value to be cached by the GV storage, I will chat with the GV team to see if they can help or if we will need to file a new ticket for this.

This is the [GV ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1712754) related to the issue above.

**Comment by Amejia481** [comment]
Telemetry follow up ticket https://github.com/mozilla-mobile/android-components/issues/10340

**Comment by Amejia481** [comment]
From the manual testing done by the QA everything looks good https://github.com/mozilla-mobile/fenix/issues/19587#issuecomment-847867305,  I think we just have to wait until the next nightly cycle for landing this.

**Review by csadilek — APPROVED** [review]
⌛ 🚢

**Pull request merged**
